### PR TITLE
Add some OpenSSL 3 QUIC raw bindings

### PR DESCRIPTION
**Pull request by bdbai**
This PR mainly focuses on `openssl-sys` and raw bindings to OpenSSL 3 QUIC features that I think might be useful for implementing high-level, ergonomic QUIC bindings in `openssl` crate. Also fixed some minor issues when using `bindgen` feature with Windows MSVC.

Related: #2242

### CODE REVIEW

File: openssl-sys/src/tls1.rs [hunk 1]
```diff
@@ -74,7 +74,16 @@ pub unsafe fn SSL_CTX_set_tlsext_servername_callback__fixed_rust(
...
-    SSL_CTX_callback_ctrl__fixed_rust(ctx, SSL_CTRL_SET_TLSEXT_SERVERNAME_CB, mem::transmute(cb))
+    SSL_CTX_callback_ctrl__fixed_rust(
+        ctx,
+        SSL_CTRL_SET_TLSEXT_SERVERNAME_CB,
+        mem::transmute::<
+            std::option::Option<
+                unsafe extern "C" fn(*mut SSL, *mut c_int, *mut libc::c_void) -> i32,
+            >,
+            std::option::Option<unsafe extern "C" fn()>,
+        >(cb),
```
Comment (sfackler) [inline openssl-sys/src/tls1.rs:85]: What is the purpose of this change?

Comment (bdbai) [reply openssl-sys/src/tls1.rs:85]: This is to address the clippy errors thrown during the first commit’s CI checks: https://github.com/sfackler/rust-openssl/actions/runs/9774772847/job/26983758164

Comment (sfackler) [reply openssl-sys/src/tls1.rs:85]: Thanks!
